server/controllers/finance/cashflow.js

```diff
@@ -0,0 +1,476 @@
+	var query =
+		'SELECT id, number, start_date, end_date ' +
+		'FROM period WHERE DATE(start_date) >= DATE(?) AND DATE(end_date) <= DATE(?)';
+	return db.exec(query, [dateFrom, dateTo]);
+}
+
+/**
+ * @function document
+ * @description process and render the cashflow report document
+ */
+function document(req, res, next) {
+  let session = {}, params = req.query;
+
+  let pageOptions = {
+    orientation: 'landscape',
+    lang: params.lang
+  };
+
+  processingCashflowReport(params)
+  .then(reporting)
+  .then(labelization)
+  .then(() => {
+    return rm.build(req, session, './server/controllers/finance/reports/cashflow.report.handlebars', pageOptions);
+  })
+  .spread((document, headers) => {
+    res.set(headers).send(document);
+  })
+  .catch(next);
+
+  /**
+   * @function reporting
+   * @param {array} rows all transactions of the given cashbox
+   * @description
+   * processing data for the report, the process is as follow :
+   * step 1. initialization : initialize all global array and objects
+   * step 2. openning balance : process for getting the openning balance
+   * step 3. grouping : group incomes and expenses by periods
+   * step 4. summarization : get all periodical openning balance
+   * step 5. labelization : define unique labels for incomes and expenses,
+   * and process all totals needed
+   * @todo: Must convert values with enterprise exchange rate
+   * @todo: Must use a date utility to format date into 'yyyy-mm-dd'
+   */
+  function reporting(rows) {
+    initialization();
+
+    session.periodicData = rows.flows;
+    /** @todo: convert into enterprise currency */
+    session.openningBalance = rows.openningBalance.balance;
+
+    session.periodicData.forEach(function (flow) {
+      groupingResult(flow.incomes, flow.expenses, util.sqlDate(flow.period.start_date));
+    });
+
+    session.periodStartArray = session.periodicData.map(function (flow) {
+      return util.sqlDate(flow.period.start_date);
+    });
+
+    /** openning balance by period */
+    session.periodicData.forEach(function (flow) {
+      summarization(util.sqlDate(flow.period.start_date));
+    });
+  }
+
+  /**
+   * @function initialization
+   * @description initialize global arrays and objects for the cashflow report
+   */
+  function initialization () {
+    session.incomes          = {};
+    session.expenses         = {};
+    session.summationIncome  = {};
+    session.summationExpense = {};
+    session.sum_incomes      = {};
+    session.sum_expense      = {};
+    session.periodicBalance  = {};
+    session.periodicOpenningBalance = {};
+    session.incomesLabels    = [];
+    session.expensesLabels   = [];
+    session.totalIncomes     = {};
+    session.totalExpenses    = {};
+  }
+
+  /**
+   * @function summarization
+   * @param {object} period An object wich reference a specific period
+   * @description process for getting openning balance for each periods
+   */
+  function summarization (period){
```

Comment:
Is it not possible to do most of this in SQL?  It would be _much, much_ more efficient.  As currently written, this will likely lock the server up for every single user if there is lots of data to process...